c/tskit/tables.c

```diff
@@ -511,6 +511,53 @@ calculate_max_rows(tsk_size_t num_rows, tsk_size_t max_rows,
     return ret;
 }
 
+static int
+calculate_max_length(tsk_size_t current_length, tsk_size_t max_length,
+    tsk_size_t max_length_increment, tsk_size_t additional_length,
+    tsk_size_t *ret_new_max_length)
+{
+    tsk_size_t new_max_length;
+    int ret = 0;
+
+    if (check_offset_overflow(current_length, additional_length)) {
+        ret = TSK_ERR_COLUMN_OVERFLOW;
+        goto out;
+    }
+
+    if (current_length + additional_length <= max_length) {
+        new_max_length = max_length;
+    } else {
+        if (max_length_increment == 0) {
+            /* Doubling by default */
+            new_max_length = TSK_MIN(max_length * 2, TSK_MAX_SIZE);
+            /* Add some constraints to prevent very small allocations */
+            if (new_max_length < 65536) {
+                new_max_length = 65536;
+            }
+            /* Prevent allocating more than 100MB additional unless needed*/
+            if (new_max_length - max_length > 104857600) {
+                new_max_length = max_length + 104857600;
+            }
+            new_max_length = TSK_MAX(new_max_length, current_length + additional_length);
```

Comment:
Maybe factor this TSK_MAX out to after the two branches? It's doing the same thing, right?